### PR TITLE
tests: Initialize test covering

### PR DESCRIPTION
**Pull request by georgebisbas**

### CODE REVIEW

**Comment by math-fehr** [comment]
Thanks! This looks indeed the right direction!

**Comment by webmiche** [comment]
I agree, this is probably a good point to start and the right direction to continue!

File: tests/test_ir.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+from xdsl.ir import Block, Region
+from xdsl.dialects.arith import Addi, Subi, Constant
+from xdsl.dialects.builtin import i32
+
+
+def test_block_build():
```
Comment (webmiche) [inline tests/test_ir.py:6]: I would describe this test more like `test_ops_accessor` or something. Even though you are building a block, you are not actually doing assertions over just the built block but over regions and results, right?

File: tests/test_ir.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    # Create a region to include a, b, c
+    region = Region.from_block_list([block0])
+
+    assert len(region.ops) == 3
+    assert len(region.blocks[0].ops) == 3
+
+    # Operation to subtract b from a
+    d = Subi.get(a, b)
+
+    assert d.results[0].__eq__(c.results[0]) is False
```
Comment (webmiche) [inline tests/test_ir.py:26]: Couldn't you use `==`, or rather `!=`, here? AFAIU, this should dispatch to that method, right?

**Review by webmiche — APPROVED** [review]
IMO, we currently have a lot of functionality in `ir.py`, so I feel it might make sense to have an entire folder `tests/ir/` that then contains files like `test_ssa`, `test_blocks`, ...

Otherwise LGTM. We can also merge and refactor later, I am fine with both!

File: tests/test_ir.py (hunk 2 above)
Comment (georgebisbas) [reply tests/test_ir.py:26]: Added == to check

File: tests/test_ir.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
+from xdsl.ir import Block, Region
+from xdsl.dialects.arith import Addi, Subi, Constant
+from xdsl.dialects.builtin import i32
+
+
+def test_block_build():
+    # This test creates two FuncOps with different approaches that
+    # represent the same code and checks their structure
+    # Create two constants and add them, add them in a region and
+    # create a function
```
[the remark this answers is not in the document]
Comment (georgebisbas) [reply tests/test_ir.py:10]: dropping yes

File: tests/test_ir.py (hunk 1 above)
Comment (georgebisbas) [reply tests/test_ir.py:6]: ok

File: tests/test_ir.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    # Create a region to include a, b, c
+    region = Region.from_block_list([block0])
+
+    assert len(region.ops) == 3
+    assert len(region.blocks[0].ops) == 3
+
+    # Operation to subtract b from a
+    d = Subi.get(a, b)
+
+    assert (d.results[0] == c.results[0]) is False
```
Comment (math-fehr) [inline tests/test_ir.py:22]: What is the difference between `(a == b) is False` and `a != b` ?
In any case, here I think checking for `a is not b` is best.

**Review by math-fehr — APPROVED** [review]
Looks good to me! I just added a comment about an assert.
Once this is resolved, we can merge it!

File: tests/test_ir.py (hunk 4 above)
Comment (georgebisbas) [reply tests/test_ir.py:22]: Thanks, I was not aware of this:
https://stackoverflow.com/questions/4352244/should-ne-be-implemented-as-the-negation-of-eq

File: tests/test_mlctx.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+@irdl_attr_definition
+class DummyAttr2(ParametrizedAttribute):
+    name = "dummy_attr2"
+
+
+def test_registration_exceptions():
+    ctx = MLContext()
+    ctx.register_op(DummyOp)
+    with pytest.raises(Exception):
```
[the remark this answers is not in the document]
Comment (georgebisbas) [reply tests/test_mlctx.py:30]: Added here, to think for a future PR: https://github.com/xdslproject/xdsl/issues/269

File: tests/test_ssavalue.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
+import pytest
+
+from xdsl.dialects.builtin import i32, StringAttr
+from xdsl.dialects.arith import Constant
+
+from xdsl.ir import Block, Operation, SSAValue
+from xdsl.irdl import irdl_op_definition, ResultDef
+
+
+def test_ssa():
```
[the remark this answers is not in the document]
Comment (georgebisbas) [reply tests/test_ssavalue.py:10]: Maybe yes, this restructuring can happen a bit later

File: tests/test_ir.py (hunk 4 above)
Comment (georgebisbas) [reply tests/test_ir.py:22]: I want to cover the __eq__ part of code

**Pull request merged**